### ISSUE TITLE
Make ~ bind tighter than +

### ISSUE THREAD

**Issue by mikeshulman** [HoTT/HoTT#1335]
The discussion at #1333 suggests that the current precedence order, which we inherited from Coq, is more or less accidental, and it might be better if `~` bound tighter than binary operators like `+` and `*`.  Let's try it!

**Comment by Alizter** [comment]
This is the file where the notations are set:
https://github.com/HoTT/HoTT/blob/master/coq/theories/Init/Notations.v

Here are some relevant lines:
```coq
Reserved Notation "~ x" (at level 75, right associativity).
Reserved Notation "x + y" (at level 50, left associativity).
```

What do we suggest the new levels be? I suppose we should change the level of ~ to something like 1. Any suggestions?

**Comment by peterlefanulumsdaine** [comment]
Level 35 sounds good to me.  I’ve tried this out in [`negation-precedence`](https://github.com/peterlefanulumsdaine/HoTT/tree/negation-precedence), and it works very well — it makes things strictly more human-readable, at least to my eye. In particular, there were a few places where the old precedence was actively misleading: e.g. in `Bool.v`, expressions like `~false = true`, which (at least to me) looks like it means `(~false) = true` (with `~` as boolean negation), but actually meant `~(false = true)`.

Regarding alternative precedences: the other big difference would be to make `~` bind tighter than function application, i.e. level <10.  That seems good when `~` is under a function, e.g. `PropHolds ~(x < y)` is nice, but bad when `~ ` is outside a function, e.g. `~ equiv p q` is a little hard to parse.

Within the precedence range between 10 (function application) and 40 (where binary operators/relations start), I see very little difference — not much to choose between e.g. 35 and 25.

**Comment by jdchristensen** [comment]
It looks to me like all of the changes in #1351 apply even if the precedences aren't changed, so to me writing things in that form would be good style, whether or not we change the precedence.  Changing the precedence would force people to use that style, which I think is more clear, but should also be balanced against having less compatibility with standard Coq.  A crazy idea would be to check that the library builds with *either* precedence, so nobody can be confused by the meaning of what we write.

**Comment by mikeshulman** [comment]
Yes, precedence 35 looks good.  I agree that everything in #1351 is an improvement.

For comparison, what are some things we could write with the new precedence that would not parse correctly with the old precedence?  Are they things we'd want to permit or prefer to forbid?

**Comment by peterlefanulumsdaine** [comment]
`~A + A` parses differently with the new and old precedences: as `(~A) + A` under the new, and `~(A+A)` under the old.  To my eye, the old parsing is definitely wrong (actively misleading), while the new parsing is good and would be the natural reading for most mathematicians.

I can’t think of any examples where the new precedence (level 35) gives an actively misleading reading, unlike the old.  I agree with what @mikeshulman said in #1333 : in maths, unary operators always bind tighter than binary, as far as I can think of.  And while I agree it’s good to be cautious about incompatibility with Coq, I find it hard to imagine how this particular one could cause any significant friction.  So I’m in favour of just changing to level 35, and endorsing the use of whatever that accepts.

On the other hand, explicitly parenthesising negations isn’t too onerous.  So if @jdchristensen or others prefer, I don’t object to either requiring that in our parsing (if that’s technically feasible), or as a more lightweight option, changing this to level 35 but adding a style guideline about this.

**Comment by mikeshulman** [comment]
I also lean towards just changing to level 35.  It's okay to be a little radical sometimes.

**Comment by jdchristensen** [comment]
I think the change is good.  If I ever have to write `~A + B`, I'll put in parentheses so it's completely unambiguous.
